lua/telescope/actions/init.lua

```diff
@@ -652,6 +652,55 @@ actions.smart_add_to_loclist = function(prompt_bufnr)
   smart_send(prompt_bufnr, 'a', 'loclist')
 end
 
+actions.hop = function(prompt_bufnr, opts)
+  opts = opts or {}
+  local escape = 27
+  local default_keys = {
+    "a", "s", "d", "f", "g", "h", "j", "k", "l", ";",
+    "q", "w", "e", "r", "t", "u", "v", "w",
+    "A", "S", "D", "F", "G", "H", "J", "K", "L", "W", "E", "R", "T", "U"}
+  opts.keys = vim.F.if_nil(opts.keys, default_keys)
+  opts.hl_group = vim.F.if_nil(opts.hl_group, "Search")
+
+  local current_picker = action_state.get_current_picker(prompt_bufnr)
+  local max_results = current_picker.max_results
+  local num_results = current_picker.manager:num_results()
+  local results_bufnr = current_picker.results_bufnr
+  local sorting_strategy = current_picker.sorting_strategy
+
+  local ns = vim.api.nvim_create_namespace "teleshopping"
+  local keyline = {}
+
+  for i = 1, math.min(num_results, max_results, #opts.keys) do
+    local key = opts.keys[i]
+    local linenr = sorting_strategy == "descending" and max_results - i or i
+    vim.api.nvim_buf_set_extmark(
+      results_bufnr,
+      ns,
+      linenr,
+      0,
+      { virt_text = { { key, opts.hl_group } }, virt_text_pos = "overlay" }
+    )
+    keyline[key] = linenr
+  end
+  -- ensure marks are drawn before getchar is executed
+  vim.cmd [[redraw]]
+
+  local char = vim.fn.getchar()
+  local key = vim.fn.nr2char(char)
+
+  -- TODO fix error of buffer previewer
```

Comment:
I think you fixed that one :) good job.

Another option would be to teardown previewer before selection and set it to nil. Then it will not try to update the previewer. I think.

Reply:
Just to avoid ambiguity, without my fix the buffer previewer would have complained about something like this in default mappings 
```lua
i = {
  ['<C-space>'] = actions.hop + actions.select_default
}
```
Since -- my hunch -- the buffer was de facto teared down before the previewer executes the `vim.schedule`d function to show it (though as a non-CS person who's never really used async I might be wrong :).

Anyways, `actions.select_default` calls `actions.close` which includes a buffer preview teardown via ultimately calling `action_set.edit` which already and  probably just "too late" (due to scheduling) does what you're referring to? Consequently, tearing down the previewer in time I'm sure would avoid this but the validity check for the buffer previewer should not induce any adverse side effects and, if at all apart from this, silently avoid other similar hurdles in the future :sweat_smile: . So I'd prefer to leave it as with the check whether the buffer is valid in `buffer_previewer.lua`. 

(`<C-space>' works for me nicely since capslock is mapped to ctrl :laughing: )

Reply:
Yeah i am fine with the current fix. Thanks :)